### PR TITLE
Role based users in Projects

### PR DESCRIPTION
**Pull request by Pearl1594**
## Description
Enabling Role Based users in projects
Primate PR related to the FR: apache/cloudstack-primate#382

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

### CODE REVIEW

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@Pearl1594 can you put the related Primate PR link in the PR description?

**Comment by Pearl1594** [comment]
Primate PR related to the FR: https://github.com/apache/cloudstack-primate/pull/382

**Comment by Pearl1594** [comment]
@rhtyd Could you please review the PR

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan test

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan package

File: api/src/main/java/org/apache/cloudstack/api/ApiConstants.java [hunk 1]
```diff
@@ -472,11 +473,16 @@
     public static final String PROJECT = "project";
     public static final String ROLE = "role";
     public static final String ROLE_ID = "roleid";
+    public static final String ACCOUNT_ROLE_TYPE = "accountroletype";
```
Comment (rhtyd) [inline api/src/main/java/org/apache/cloudstack/api/ApiConstants.java:476]: roletype is enough?

File: api/src/main/java/org/apache/cloudstack/api/ApiConstants.java [hunk 2]
```diff
@@ -472,11 +473,16 @@
...
+    public static final String PROJECT_ROLE_ID = "projectroleid";
+    public static final String PROJECT_ROLE_NAME = "projectrolename";
+    public static final String PROJECT_ROLE_TYPE = "projectroletype";
     public static final String ROLE_TYPE = "roletype";
     public static final String ROLE_NAME = "rolename";
     public static final String PERMISSION = "permission";
     public static final String RULE = "rule";
     public static final String RULE_ID = "ruleid";
+    public static final String PROJECT_RULE_ID = "projectruleid";
```
Comment (rhtyd) [inline api/src/main/java/org/apache/cloudstack/api/ApiConstants.java:485]: nit - probably this means projectrolepermissionid? (yes, naming is hard, I'm trying to guess what is parameter is for)

File: api/src/main/java/org/apache/cloudstack/api/BaseCmd.java [hunk 3]
```diff
@@ -264,6 +269,10 @@ public String getActualCommandName() {
      */
     public abstract long getEntityOwnerId();
 
+    public List<Long> getEntityOwnerIds() {
+        return null;
```
Comment (rhtyd) [inline api/src/main/java/org/apache/cloudstack/api/BaseCmd.java:273]: Could this cause any exception; maybe return empty array instead or null?

File: engine/schema/src/main/java/org/apache/cloudstack/acl/dao/ProjectRoleDaoImpl.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        ProjectRoleSearch = createSearchBuilder();
+        ProjectRoleSearch.and("name", ProjectRoleSearch.entity().getName(), SearchCriteria.Op.LIKE);
+        ProjectRoleSearch.and("project_id", ProjectRoleSearch.entity().getProjectId(), SearchCriteria.Op.EQ);
+        ProjectRoleSearch.done();
+
+    }
+    @Override
+    public List<ProjectRoleVO> findByName(String name, Long projectId) {
+        SearchCriteria<ProjectRoleVO> sc = ProjectRoleSearch.create();
+        sc.setParameters("name", "%" + name + "%");
```
Comment (rhtyd) [inline engine/schema/src/main/java/org/apache/cloudstack/acl/dao/ProjectRoleDaoImpl.java:43]: @Pearl1594 can you add a Strings.isNullOrEmpty here

File: api/src/main/java/org/apache/cloudstack/api/BaseCmd.java (hunk 3 above)
Comment (Pearl1594) [reply api/src/main/java/org/apache/cloudstack/api/BaseCmd.java:273]: the only place referenced - ParamProcessWorker class has been handled to check if a null is returned

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan  package

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@Pearl1594 can you merge latest master and rekick tests?

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan test

File: server/src/main/java/org/apache/cloudstack/acl/ProjectRoleManagerImpl.java [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+            public Boolean doInTransaction(TransactionStatus status) {
+                List<? extends ProjectRolePermission> rolePermissions = projRolePermissionsDao.findAllByRoleIdSorted(role.getId(), projectId);
+                if (rolePermissions != null && !rolePermissions.isEmpty()) {
+                    for (ProjectRolePermission rolePermission : rolePermissions) {
+                        projRolePermissionsDao.remove(rolePermission.getId());
+                    }
+                }
+                if (projRoleDao.remove(role.getId())) {
+                    ProjectRoleVO projRoleVO = projRoleDao.findByIdIncludingRemoved(role.getId());
+                    projRoleVO.setName(null);
```
Comment (nvazquez) [inline server/src/main/java/org/apache/cloudstack/acl/ProjectRoleManagerImpl.java:271]: Why is this required after deletion?

**Comment by nvazquez** [comment]
Looks good @Pearl1594 thanks, left some comments

**Comment by Pearl1594** [comment]
@blueorangutan package

File: api/src/main/java/org/apache/cloudstack/api/command/user/project/UpdateProjectCmd.java [hunk 6]
```diff
@@ -68,11 +89,49 @@ public String getDisplayText() {
...
+        return accountId;
+    }
+
     @Override
     public String getCommandName() {
         return s_name;
     }
 
+    public Boolean isSwapOwner() {
+        if (swapOwner != null) {
```
[the remark this answers is not in the document]
Comment (Pearl1594) [reply api/src/main/java/org/apache/cloudstack/api/command/user/project/UpdateProjectCmd.java]: done

File: api/src/main/java/org/apache/cloudstack/api/command/user/account/AddAccountToProjectCmd.java [hunk 7]
```diff
@@ -72,6 +87,21 @@ public String getEmail() {
         return email;
     }
 
+    public Long getProjectRoleId() {
+        return projectRoleId;
+    }
+
+    public ProjectAccount.Role getRoleType() {
+        if (roleType != null) {
```
[the remark this answers is not in the document]
Comment (Pearl1594) [reply api/src/main/java/org/apache/cloudstack/api/command/user/account/AddAccountToProjectCmd.java]: done

File: api/src/main/java/org/apache/cloudstack/api/command/user/account/AddUserToProjectCmd.java [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+    }
+
+    public String getEmail() { return email; }
+
+    public Long getProjectRoleId() {
+        return projectRoleId;
+    }
+
+    public ProjectAccount.Role getRoleType() {
+        if (roleType != null) {
```
[the remark this answers is not in the document]
Comment (Pearl1594) [reply api/src/main/java/org/apache/cloudstack/api/command/user/account/AddUserToProjectCmd.java]: done

File: api/src/main/java/org/apache/cloudstack/api/command/user/project/UpdateProjectCmd.java [hunk 9]
```diff
@@ -68,11 +89,49 @@ public String getDisplayText() {
...
+    public ProjectAccount.Role getRoleType(String role) {
+        String type = role.substring(0, 1).toUpperCase() + role.substring(1).toLowerCase();
+        if (!EnumUtils.isValidEnum(ProjectAccount.Role.class, type)) {
+            throw new InvalidParameterValueException("Only Admin or Regular project role types are valid");
+        }
+        return Enum.valueOf(ProjectAccount.Role.class, type);
+    }
+
+    public ProjectAccount.Role getAccountRole() {
+        if (roleType != null) {
```
[the remark this answers is not in the document]
Comment (Pearl1594) [reply api/src/main/java/org/apache/cloudstack/api/command/user/project/UpdateProjectCmd.java]: done

File: api/src/main/java/org/apache/cloudstack/api/command/user/project/UpdateProjectCmd.java [hunk 10]
```diff
@@ -84,14 +143,38 @@ public long getEntityOwnerId() {
...
+
     /////////////////////////////////////////////////////
     /////////////// API Implementation///////////////////
     /////////////////////////////////////////////////////
 
     @Override
     public void execute() throws ResourceAllocationException {
         CallContext.current().setEventDetails("Project id: " + getId());
-        Project project = _projectService.updateProject(getId(), getDisplayText(), getAccountName());
+        if (getAccountName() != null && (getUserId() != null)) {
```
[the remark this answers is not in the document]
Comment (Pearl1594) [reply api/src/main/java/org/apache/cloudstack/api/command/user/project/UpdateProjectCmd.java]: done

File: engine/schema/src/main/java/org/apache/cloudstack/acl/ProjectRolePermissionVO.java [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+    @Column(name = "project_id")
+    private long projectId;
+
+    @Column(name = "project_role_id")
+    private long projectRoleId;
+
+    @Column(name = "sort_order")
+    private long sortOrder = 0;
+
+    public ProjectRolePermissionVO() { super(); }
```
[the remark this answers is not in the document]
Comment (Pearl1594) [reply engine/schema/src/main/java/org/apache/cloudstack/acl/ProjectRolePermissionVO.java]: done

File: server/src/main/java/org/apache/cloudstack/acl/ProjectRoleManagerImpl.java (hunk 5 above)
Comment (Pearl1594) [reply server/src/main/java/org/apache/cloudstack/acl/ProjectRoleManagerImpl.java:271]: Only to clear the name pf the role from the DB once its been removed. May not be required, but followed what was done for Roles

**Comment by Pearl1594** [comment]
Thanks @nvazquez , I've addressed the comments

**Comment by Pearl1594** [comment]
@blueorangutan test

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan test

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan test

File: server/src/main/java/org/apache/cloudstack/acl/ProjectRoleManagerImpl.java (hunk 5 above)
Comment (nvazquez) [reply server/src/main/java/org/apache/cloudstack/acl/ProjectRoleManagerImpl.java:278]: I'm ok with leaving the update for consistency

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan test

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan test

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan test

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan test

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan test

**Comment by Pearl1594** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
Rekicked travis

File: results.xml [hunk 12]
```diff
@@ -0,0 +1,3 @@
+<?xml version="1.0" encoding="UTF-8"?><testsuite name="nosetests" tests="3" errors="0" failures="0" skip="0"><testcase classname="integration.smoke.test_enable_role_based_users_in_projects.TestRoleBasedUsersInProjects" name="test_add_account_to_project_with_project_role" time="3.046"></testcase><testcase classname="integration.smoke.test_enable_role_based_users_in_projects.TestRoleBasedUsersInProjects" name="test_add_multiple_admins_in_project" time="6.491"><system-out><![CDATA[pid == 3a16742f-819b-4d44-89d9-1f4122602038
+2
+]]></system-out></testcase><testcase classname="integration.smoke.test_enable_role_based_users_in_projects.TestRoleBasedUsersInProjects" name="test_add_user_to_project_with_project_role" time="2.713"></testcase></testsuite>
```
Comment (DaanHoogland) [inline results.xml:3]: this file is missing a license. either add it to the ignore list or add a license.

Comment (DaanHoogland) [reply results.xml]: I see this results.xml is changed now but still without license. I think it makes sense to either remove it from git or add it to the exclusion list for the rat, @Pearl1594 .

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
test env failed restarting
@blueorangutan test

**Comment by rhtyd** [comment]
Test LGTM, failure in CKS are env related
